File: frontend/lib/api/virtual-sensors.ts
import type { VirtualSensorConfiguration, WearableSensorWindow } from "@/lib/types/virtual-sensors";

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";
export type SensorWindowRequest = { motion_condition: "STABLE_REST" | "ACTIVE_MOTION" | "POSTURE_TRANSITION"; contact_condition: "NOMINAL_CONTACT" | "UNSTABLE_CONTACT" | "POOR_CONTACT"; temperature_offset_c: number; };

async function parse<T>(response: Response): Promise<T> { if (!response.ok) { const body = await response.json().catch(() => null) as { detail?: { message?: string } | string } | null; throw new Error(typeof body?.detail === "string" ? body.detail : body?.detail?.message ?? `Virtual sensor request failed with HTTP ${response.status}`); } return response.json() as Promise<T>; }
export async function fetchVirtualSensorConfig(signal?: AbortSignal): Promise<VirtualSensorConfiguration> { return parse(await fetch(`${API_BASE_URL}/api/v1/virtual-sensors/config`, { cache: "no-store", signal })); }
export async function acquireSensorWindow(request: SensorWindowRequest): Promise<WearableSensorWindow> { return parse(await fetch(`${API_BASE_URL}/api/v1/virtual-sensors/window`, { method: "POST", headers: { "Content-Type": "application/json" }, body: JSON.stringify(request) })); }

File: backend/app/api/routes/virtual_sensors.py
from fastapi import APIRouter, HTTPException, status

from app.simulation.sensors.models import SensorWindowRequest, VirtualSensorConfiguration, WearableSensorWindow
from app.simulation.sensors.service import VirtualSensorsUnavailableError, wearable_sensor_service

router = APIRouter(prefix="/api/v1/virtual-sensors", tags=["virtual-sensors"])

@router.get("/config", response_model=VirtualSensorConfiguration)
async def get_virtual_sensor_configuration() -> VirtualSensorConfiguration:
    return wearable_sensor_service.configuration()

@router.post("/window", response_model=WearableSensorWindow)
async def acquire_sensor_window(request: SensorWindowRequest) -> WearableSensorWindow:
    try:
        return wearable_sensor_service.acquire(request)
    except VirtualSensorsUnavailableError as error:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail={"code": "VIRTUAL_SENSORS_UNAVAILABLE", "message": str(error)}) from error
